dialogue-core/src/main/java/com/palantir/dialogue/core/QueuedChannel.java

```diff
@@ -144,16 +144,33 @@ private boolean scheduleNextTask() {
         if (components == null) {
             return false;
         }
-
+        SettableFuture<Response> queuedResponse = components.response();
+        // If the future has been completed (most likely via cancel) the call should not be queued.
+        // There's a race where cancel may be invoked between this check and execution, but the scheduled
+        // request will be quickly cancelled in that case.
+        if (queuedResponse.isDone()) {
+            return true;
+        }
         try (CloseableSpan ignored = components.span().childSpan("Dialogue-request-scheduled")) {
-            Optional<ListenableFuture<Response>> response =
+            Optional<ListenableFuture<Response>> maybeResponse =
                     delegate.maybeExecute(components.endpoint(), components.request());
 
-            if (response.isPresent()) {
+            if (maybeResponse.isPresent()) {
+                ListenableFuture<Response> response = maybeResponse.get();
                 components.span().complete();
                 numRunningRequests.incrementAndGet();
-                response.get().addListener(numRunningRequests::decrementAndGet, DIRECT);
-                Futures.addCallback(response.get(), new ForwardAndSchedule(components.response()), DIRECT);
+                response.addListener(numRunningRequests::decrementAndGet, DIRECT);
+                Futures.addCallback(response, new ForwardAndSchedule(queuedResponse), DIRECT);
+                queuedResponse.addListener(
+                        () -> {
+                            if (queuedResponse.isCancelled()) {
+                                // TODO(ckozak): Consider capturing the argument value provided to cancel to propagate
+                                // here.
+                                // Currently cancel(false) will be converted to cancel(true)
+                                response.cancel(true);
```

Comment:
the javadoc on this says `@return {@code false} if the task could not be cancelled, typically because it has already completed normally;`... i wonder if it's worth putting a debug log line in that case? I feel like the timelock folks had so much time debugging cancellations that they'd want every bit of signal they could get

Reply:
good call!